reframe/utility/__init__.py

```diff
@@ -1,5 +1,136 @@
+    def global_scope_mark(self):
+        return self._global_scope
+
+    def update(self, other):
+        if not isinstance(other, collections.abc.Mapping):
+            raise TypeError('ScopedDict may only be initialized '
+                            'from a mapping type')
+
+        for scope, scope_dict in other.items():
+            self._check_scope_type(scope, scope_dict)
+            self.data.setdefault(scope, {})
+            for k, v in scope_dict.items():
+                self.data[scope][k] = v
+
+    def __str__(self):
+        # just return the internal dictionary
+        return str(self.data)
+
+    def _check_scope_type(self, key, value):
+        if not isinstance(key, str):
+            raise TypeError('scope keys in a ScopedDict must be strings')
+
+        if not isinstance(value, collections.abc.Mapping):
+            raise TypeError('scope namespaces must be mappings')
+
+        for k in value.keys():
+            if not isinstance(k, str):
+                raise TypeError('keys must be strings')
+
+    def _keyinfo(self, key):
+        key_parts = key.rsplit(self._scope_sep, maxsplit=1)
+        if len(key_parts) == 2:
+            return (key_parts[0], key_parts[1])
+        else:
+            return (self._global_scope, key_parts[0])
+
+    def _parent_scope(self, scope):
+        scope_parts = scope.rsplit(':', maxsplit=1)[:-1]
+        return scope_parts[0] if scope_parts else self._global_scope
+
+    def _lookup(self, key):
+        scope, lookup_key = self._keyinfo(key)
+        while scope != self._global_scope:
+            if scope in self.data and lookup_key in self.data[scope]:
+                return self.data[scope][lookup_key]
+
+            scope = self._parent_scope(scope)
+
+        # last chance to find the key
+        if scope in self.data and lookup_key in self.data[scope]:
+            return self.data[scope][lookup_key]
+
+        raise KeyError(str(key))
+
+    def __iter__(self):
+        for scope, scope_dict in self.data.items():
+            for k in scope_dict.keys():
+                yield self._scope_sep.join([scope, k])
+
+    def __contains__(self, key):
+        try:
+            self._lookup(key)
+        except KeyError:
+            return False
+        else:
+            return True
+
+    def __getitem__(self, key):
+        try:
+            return self._lookup(key)
+        except KeyError:
+            return self.__missing__(key)
+
+    def __setitem__(self, key, value):
+        scope, lookup_key = self._keyinfo(key)
+        if scope not in self.data:
+            # create the scope if does not exist
+            self.data[scope] = {}
+
+        self.data[scope][lookup_key] = value
+
+    def __delitem__(self, key):
+        """Deletes either a key or a scope if key refers to a scope.
+
+        If key refers to scope, the whole scope entry will be deleted. If not,
+        the exact key requested will be deleted. No key resolution will be
+        performed."""
```

Comment:
We should expand the description to cover the case when a key and a scope have the same name. I suggest the following:

```
If the key refers to both a scope and a key, the key will be deleted.
```